### PR TITLE
[SmartContracts] Refactor Block Assemblers Part #2

### PR DESCRIPTION
**Pull request by fassadlr**
In this part we remove the creation of the block assemblers from PowMining and PosMinting and rather inject an singleton instance into the constructor of each.

### CODE REVIEW

**Comment by fassadlr** [comment]
All unit and integration tests pass. I have synced from test and mainnet and was able to still stake coins on testnet.

**Comment by Aprogiena** [comment]
> All unit and integration tests pass. I have synced from test and mainnet and was able to still stake coins on testnet.

Excellent, there is how I think it should be that the developer always that these tests. But I would also like the reviewer to repeat them and confirm.

File: src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs [hunk 1]
```diff
@@ -21,27 +21,25 @@ public class PosBlockAssembler : PowBlockAssembler
...
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
             this.stakeChain = stakeChain;
             this.stakeValidator = stakeValidator;
         }
 
-        public override BlockTemplate CreateNewBlock(Script scriptPubKeyIn, bool fMineWitnessTx = true)
+        public override BlockTemplate Build(ChainedBlock chainTip, Script minerAddress)
         {
-            this.logger.LogTrace("({0}.{1}:{2},{3}:{4})", nameof(scriptPubKeyIn), nameof(scriptPubKeyIn.Length), scriptPubKeyIn.Length, nameof(fMineWitnessTx), fMineWitnessTx);
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(chainTip), chainTip.Height, nameof(minerAddress), nameof(minerAddress.Length), minerAddress.Length);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs]: When you log chained block, just use `{0}:'{1}'` and `nameof(chainTip), chainTip`, not just its height

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs [hunk 2]
```diff
@@ -201,55 +212,39 @@ public int Compare(TxMempoolEntry a, TxMempoolEntry b)
...
         }
 
-        /** Construct a new block template with coinbase to scriptPubKeyIn */
-
-        public override BlockTemplate CreateNewBlock(Script scriptPubKeyIn, bool fMineWitnessTx = true)
+        /// <inheritdoc/>
+        public override BlockTemplate Build(ChainedBlock chainTip, Script minerAddress)
         {
-            this.logger.LogTrace("({0}.{1}:{2},{3}:{4})", nameof(scriptPubKeyIn), nameof(scriptPubKeyIn.Length), scriptPubKeyIn.Length, nameof(fMineWitnessTx), fMineWitnessTx);
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(chainTip), chainTip.Height, nameof(minerAddress), nameof(minerAddress.Length), minerAddress.Length);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: See https://github.com/stratisproject/StratisBitcoinFullNode/pull/1361/files#r185450528

File: src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs]: Fixed 👍

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs [hunk 3]
```diff
@@ -14,11 +14,18 @@ namespace Stratis.Bitcoin.Features.Miner
 {
     public abstract class BlockAssembler
     {
-        /// <summary>Tip of the chain that this instance will work with without touching any shared chain resources.</summary>
-        /// <remarks>Using a fixed value prevents race conditions in the methods of derived classes.</remarks>
+        /// <summary>
+        /// Constructs a block template which will be passed to consensus. 
+        /// </summary>
+        /// <param name="chainTip">Tip of the chain that this instance will work with without touching any shared chain resources.</param>
+        /// <param name="minerAddress">The miner's address.</param>
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Yes you are right, I have renamed back and fixed the comment :)

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Fixed

**Comment by fassadlr** [comment]
@Aprogiena Ive re-run all tests and integration tests. All good :)

**Pull request merged**